util/src/main/java/tech/pegasys/artemis/util/bls/BLSVerify.java

```diff
@@ -13,15 +13,253 @@
+    long z2 = z1z2[1].getValue();
+
+    UInt381 x1 = UInt381.valueOf(z2 % (long) Math.pow(2, 381));
+    boolean a_flag1 = z1 % Math.pow(2, 382) / Math.pow(2, 381) == 1;
+    boolean b_flag1 = z1 % Math.pow(2, 383) / Math.pow(2, 382) == 1;
+    boolean c_flag1 = z1 % Math.pow(2, 384) / Math.pow(2, 383) == 1;
+
+    UInt381 x2 = UInt381.valueOf(z2 % (long) Math.pow(2, 381));
+    boolean a_flag2 = z2 % Math.pow(2, 382) / Math.pow(2, 381) == 1;
+    boolean b_flag2 = z2 % Math.pow(2, 383) / Math.pow(2, 382) == 1;
+    boolean c_flag2 = z2 % Math.pow(2, 384) / Math.pow(2, 383) == 1;
+
+    assert x1.getValue() < q.getValue() && x2.getValue() < q.getValue();
+    assert a_flag2 == b_flag2 && b_flag2 == c_flag2 && !c_flag2 && c_flag1;
+
+    if (b_flag1) {
+      assert !a_flag1 && x1.getValue() == x2.getValue() && x2 == UInt381.ZERO;
+      // (z1, z2) represents the point at infinity
+    } else {
+      assert z1 == x1 * i + x2 && z2 == y;
+      // then (z1, z2) represents the point (x1 * i + x2, y)
+      // where y is the valid coordinate such that the imaginary part y_im of y satisfies
+      assert y_im * 2 / q == a_flag1;
+    }
+
+    return true;
+  }
+
+  /**
+   *
+   * @param message
+   * @param domain
+   * @return
+   */
+  private UInt384[] hash_to_G2(Bytes32 message, int domain) {
+    // Initial candidate x coordinate
+    BytesValue x_re_arg = intToBytes32(domain) + (byte) 1 + message;
+    BytesValue x_im_arg = intToBytes32(domain) + (byte) 2 + message;
+    int x_re = hashToInt(Hash.hash(x_re_arg));
+    int x_im = hashToInt(Hash.hash(x_im_arg));
+    int x_coordinate = Fq2(new int[]{x_re, x_im});  // x = x_re + i * x_im;
+
+    //  Test candidate y coordinates until a one is found
+    while (true) {
+      int y_coordinate_squared = (int) Math.pow(x_coordinate, 3) +
+          Fq2(new int[]{4 ,4}); // The curve is y^2 = x^3 + 4(i + 1);
+      int y_coordinate = modular_squareroot(y_coordinate_squared);
+      if (y_coordinate != -1) { // Check if quadratic residue found
+        return multiply_in_G2(new int[]{x_coordinate, y_coordinate}, G2_cofactor);
+      }
+      x_coordinate += Fq2(new int[]{1, 0});  // Add 1 and try again
+    }
+  }
+
+  /**
+   * Convert integer to Bytes32.
+   * @param num
+   * @return
+   */
+  private Bytes32 intToBytes32(int num) {
+    // TODO
+    return Bytes32.TRUE;
+  }
+
+  /**
+   * Convert hash to int value.
+   * @param hash
+   * @return
+   */
+  private int hashToInt(Hash hash) {
+    // TODO
+    return 0;
+  }
+
+  /**
+   *
+   * @param point
+   * @param cofactor
+   * @return
+   */
+  private UInt384[] multiply_in_G2(int[] point, UInt381 cofactor) {
+    // TODO
+    return new UInt384[]{};
+  }
+
+  /**
+   * If there are two solutions the one with higher imaginary component is favored; if both solutions have equal
+   * imaginary component the one with higher real component is favored.
+   * @param value
+   * @return
+   */
+  private int modular_squareroot(int value) {
```

Comment:
TODO: What is `eighth_roots_of_unity`? Verify that this modular_squareroot works like other formulas. @jrhea @schroedingerscode @dangerousfood

Reply:
I think we need to look for a native C or C++ implementation of the pairing function e() and create a JNI wrapper for it.